lib/rule-translation/src/RuleTranslation.hs

```diff
@@ -0,0 +1,449 @@
+    FApp (AC Union)    ts                           -> printAC "union" ts
+    FApp (AC Xor)      ts                           -> printAC "xor" ts
+    FApp (NoEq (f, _)) ts | (BC.unpack f == "pair") -> printPair ts
+    FApp (NoEq (f, _)) ts                           -> (showFunction $ BC.unpack f) ++ printList ts
+    FApp (C EMap)      ts                           -> "em" ++ printList ts
+    FApp List          ts                           -> printList ts
+    where
+      printList ts = "(" ++ (intercalate ", " $ map (printTerm vars checkEq) ts) ++ ")"
+      printPair [t1,t2] = "(" ++ printTerm vars checkEq t1 ++ ", " ++ printTerm vars checkEq t2 ++ ")"
+      printAC op [t1,t2] = op ++ "(" ++ printTerm vars checkEq t1 ++ ", " ++ printTerm vars checkEq t2 ++ ")"
+      printAC op (t:ts) = op ++ "(" ++ printTerm vars checkEq t ++ ", " ++ printAC op ts ++ ")"
+
+printTerm2 :: (Show l) => Term l -> String
+printTerm2 t = case viewTerm t of
+    Lit l                                           -> showAtom2 $ show l
+    FApp (AC Mult)     ts                           -> printAC "mult" ts
+    FApp (AC Union)    ts                           -> printAC "union" ts
+    FApp (AC Xor)      ts                           -> printAC "xor" ts
+    FApp (NoEq (f, _)) ts | (BC.unpack f == "pair") -> printPair ts
+    FApp (NoEq (f, _)) ts                           -> (showFunction $ BC.unpack f) ++ printList ts
+    FApp (C EMap)      ts                           -> "em" ++ printList ts
+    FApp List          ts                           -> printList ts
+    where
+      printList ts = "(" ++ (intercalate ", " $ map printTerm2 ts) ++ ")"
+      printPair [t1,t2] = "(" ++ printTerm2 t1 ++ ", " ++ printTerm2 t2 ++ ")"
+      printAC op [t1,t2] = op ++ "(" ++ printTerm2 t1 ++ ", " ++ printTerm2 t2 ++ ")"
+      printAC op (t:ts) = op ++ "(" ++ printTerm2 t ++ ", " ++ printAC op ts ++ ")"
+
+translatePatternTerm :: (Document d, Show l) => S.Set String -> M.Map String String -> Term l -> (d, M.Map String String)
+translatePatternTerm vars helperVars t = case viewTerm t of
+    Lit l | (S.member (show l) vars || head (show l) == '\'') -> ((text "=" <> (text . showAtom $ show l)), helperVars)
+    Lit l                                                     -> ((text . showAtom $ show l), helperVars)
+    _                                                         -> (varDoc, newHelperVars)
+                                                                 where
+                                                                   (newVar, newHelperVars) = makeVariable t helperVars
+                                                                   varDoc = text newVar
+
+makeDestructorDefinition :: (Show l) => Term l -> String
+makeDestructorDefinition t = 
+  "forall " ++ intercalate ", " (map (++":bitstring") atoms) ++ ";#" ++ printTerm2 t
+  where
+    atoms = map showAtom2 . S.toList . S.fromList $ getAtoms t
```

Comment:
` S.toList . S.fromList ` this is the identity function, isn't it?

Reply:
I don't think so? I used this for duplicate elimination. But maybe there is a better way of doing this, I don't know.